File: 3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.py
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        minL, minW, ans = 3000, 3000, 3000
        n = len(landStartTime)
        m = len(waterStartTime)

        for i in range(n):
            minL = min(minL, landStartTime[i] + landDuration[i])

        for i in range(m):
            minW = min(minW, waterStartTime[i] + waterDuration[i])
            ans = min(ans, max(minL, waterStartTime[i]) + waterDuration[i])

        for i in range(n):
            ans = min(ans, max(minW, landStartTime[i]) + landDuration[i])

        return ans

File: 3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.go
func earliestFinishTime(landStartTime []int, landDuration []int, waterStartTime []int, waterDuration []int) int {
	n := len(landStartTime)
	m := len(waterStartTime)

	ans := int(^uint(0) >> 1)

	for i := 0; i < n; i++ {
		landEnd := landStartTime[i] + landDuration[i]

		for j := 0; j < m; j++ {
			startWater := max(landEnd, waterStartTime[j])
			finish := startWater + waterDuration[j]
			if finish < ans {
				ans = finish
			}
		}
	}

	for j := 0; j < m; j++ {
		waterEnd := waterStartTime[j] + waterDuration[j]

		for i := 0; i < n; i++ {
			startLand := max(waterEnd, landStartTime[i])
			finish := startLand + landDuration[i]
			if finish < ans {
				ans = finish
			}
		}
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}